### PR TITLE
Add PowerShell72 Runtime Support in Azure Automation

### PR DESCRIPTION
**Pull request by sushil490023**
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [ ] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ X] Other, please clarify:
    - We are adding new Value to existing ENUM ( Runbook Type) and corresponding new API for new Added RunbookType

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ X] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [X ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by sushil490023** [comment]
We are supporting new Runbook Type **PowerShell72** we want to add this runbook type to existing stable api version.
Also we are introducing new API route to support new Runbook type curd operation.

Breaking changes are failing because of above mentioned reason 
image

Also for Lint diff errors are failing, fixing these errors may cause definition of existing to change which may be breaking scenario for customer.

image


Can you please guide what can be done here?

**Comment by sushil490023** [comment]
Created breaking changes workitem [Scenario 25342538](https://msazure.visualstudio.com/One/_workitems/edit/25342538): Adding new Value to Enum(Runbook Type)- Breaking change Approval

**Comment by sushil490023** [comment]
Hi 
For breaking changes https://github.com/Azure/azure-rest-api-specs/pull/26057/checks?check_run_id=17598290426
image

In new API Version we are making **name** property of Resource **Module** as read only as it cannot be changed. In pervious version this check was not evaluated.

**Comment by sushil490023** [comment]
Hi for Lint Diff error failing : https://github.com/Azure/azure-rest-api-specs/pull/26057/checks?check_run_id=17598290751

image

 For Module resource, we need file uploads (currently taking a short lived blob content uri) in PUT request, so couldn't make PUT contract same as GET contract.
 PUT has one extra property for contentURI and which we dont return as part of GET response.

File: specification/automation/resource-manager/Microsoft.Automation/common/v5/definitions.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      }
+    }
+  },
+  "info": {
+    "version": "2023-11-01",
+    "title": "Common types"
+  },
+  "paths": {},
+  "definitions": {
+    "PrivateEndpointConnection": {
```
Comment (pcsin) [inline specification/automation/resource-manager/Microsoft.Automation/common/v5/definitions.json:27]: >PrivateEndpointConnection [](http://example.com/codeflow?start=5&length=25)

the object seem similar to the one defined in `common-types` here - [https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/privatelinks.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/privatelinks.json)

Is there a specific reason to not use it?

Comment (mahakjain314) [reply specification/automation/resource-manager/Microsoft.Automation/common/v5/definitions.json]: Removed these Private Link related definitions from the common definitions.json as private links APIs are not part of this stable version.

**Comment by pcsin** [comment]
>                   },

nit: better to move under `common` types?

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:65 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by mahakjain314** [comment]
> > ```
> >                   "$ref": "../../common/v1/definitions.json#/parameters/ResourceGroupNameParameter"
> > ```
> 
> pls use known params from the ones defined in `common-types`. Same for `subscription`, `apiversion` params, and `errorResponse` object types - [https://github.com/sushil490023/azure-rest-api-specs/blob/AddPS72/specification/common-types/resource-management/v5/types.json`](https://github.com/sushil490023/azure-rest-api-specs/blob/AddPS72/specification/common-types/resource-management/v5/types.json`)
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:53 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Updated all the references in the file to use from common-types, as suggested.

**Comment by pcsin** [comment]
>               }

would be better to refer from Proxy resource as it's base type from common types.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:590 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by mahakjain314** [comment]
> > ```
> >               },
> > ```
> 
> nit: better to move under `common` types?
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:65 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Moved ModuleName parameter to common-type

**Comment by pcsin** [comment]
>                           "$ref": "#/definitions/ModuleCreateOrUpdateParameters"

I will have to check if different request, and response in a PUT semantic would be fine for this case, as I read it is specific to your scenario.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:171 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by pcsin** [comment]
>               }

similar to the comment for PUT, better to refer base types.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:709 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by pcsin** [comment]
>               "contentLink": {

is `contentLink` not a `required` property?

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:678 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by mahakjain314** [comment]
> > ```
> >           "contentLink": {
> > ```
> 
> is `contentLink` not a `required` property?
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:678 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

ContentLink is required for Create but not required for Update. That is the only difference between ModuleCreateOrUpdateParameters and ModuleUpdateParameters

**Comment by mahakjain314** [comment]
> > ```
> >               "$ref": "#/definitions/ModuleUpdateProperties",
> > ```
> 
> curious, can `ModuleCreateOrUpdateParameters` be used here? from the looks of it `ModuleUpdateParameters` and `ModuleCreateOrUpdateParameters` look similar
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:689 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Similar explanation as above. ContentLink is required in only one.

**Comment by mahakjain314** [comment]
> > ```
> >               "type": "boolean",
> > ```
> 
> is modelling this boolean possible to model as an enum similar to `provisioningState` to be future proof in case of addition on new states?
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:601 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

We cannot change the type for this resource because we have several other resources which adhere to the same contract - module.json, python2package.json and python3package.json. Changing this will cause disparity.

**Comment by pcsin** [comment]
>       "ActivityListResult": {

these types are not used in for the current resource type

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:471 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

**Comment by mahakjain314** [comment]
> > ```
> >           }
> > ```
> 
> would be better to refer from Proxy resource as it's base -type from common types
> 
> https://github.com/sushil490023/azure-rest-api-specs/blob/8b21321c98b0136771019f5804d100589e92a5a4/specification/common-types/resource-management/v5/types.json#L9C5-L9C5
> 
> Using base types will avoid other issues such as `name`, and `location` being `readOnly`
> 
> Optionally can include `tags`
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:590 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Location doesnt  seems to be part ProxyResource
Reference - https://github.com/sushil490023/azure-rest-api-specs/blob/8b21321c98b0136771019f5804d100589e92a5a4/specification/common-types/resource-management/v5/types.json#L93C1-L94C1

image

File: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      "body": {
+        "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Automation/automationAccounts/myAutomationAccount33/powershell72modules/OmsCompositeResources",
+        "name": "OmsCompositeResources",
+        "type": "Microsoft.Automation/AutomationAccounts/PowerShell72Modules",
+        "location": "East US 2",
+        "tags": {},
+        "etag": null,
+        "properties": {
+          "isGlobal": false,
+          "version": null,
```
Comment (pcsin) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json:33]: >null [](http://example.com/codeflow?start=21&length=4)

the response property (version) is not going with the input request. Pls fix.

**Comment by pcsin** [comment]
>       }

nit: these two objects types are not used in the current resource type defintion. Could be removed?

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:758 in a4409fe. [](commit_id = a4409fe62d1f8e3275929c15c9ea93a481de5533, deletion_comment = False)

File: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/listPowerShell72ModulesByAutomationAccount.json [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+              "lastModifiedTime": "2017-03-17T16:57:48.343+00:00",
+              "provisioningState": "Created"
+            }
+          },
+          {
+            "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Automation/automationAccounts/myAutomationAccount33/powershell72modules/Azure.Storage",
+            "name": "Azure.Storage",
+            "properties": {
+              "isGlobal": true,
+              "version": null,
```
[the remark this answers is not in the document]
Comment (mahakjain314) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/listPowerShell72ModulesByAutomationAccount.json]: Fixed the examples. Although, version can still be null while the operation is in progress or fails.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          "sizeInBytes": 0,
+          "activityCount": 0,
+          "creationTime": "2017-03-29T15:41:47.003+00:00",
+          "lastModifiedTime": "2017-03-29T15:42:10.567+00:00",
+          "error": {
+            "code": null,
+            "message": null
+          },
+          "provisioningState": "Creating",
+          "isComposite": false
```
[the remark this answers is not in the document]
Comment (mahakjain314) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json:68]: That's by design. ContentLink is generally a short-lived blob location to provide the content. We do not intend to store that or return back as part of output.

**Comment by mahakjain314** [comment]
> > ```
> >   }
> > ```
> 
> nit: these two objects types are not used in the current resource type defintion. Could be removed?
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:758 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Removed these definitions. 

> > ```
> >   "ActivityListResult": {
> > ```
> 
> this object types (and its hierarchy) is not used in for the current resource type (powershell72modules), do we need it?
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:471 in [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533). [](commit_id = [a4409fe](https://github.com/Azure/azure-rest-api-specs/commit/a4409fe62d1f8e3275929c15c9ea93a481de5533), deletion_comment = False)

Removed the unused definitions.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json (hunk 2 above)
Comment (mahakjain314) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/examples/createOrUpdatePowerShell72Module.json]: Fixed.

**Comment by pcsin** [comment]
>         },

can we make `name` and `location` readOnly?

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:394 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>           "description": "Gets or sets the module update properties."

nit: pls fix documentation. I would expect this to be only updating module properties, hence `sets` and not `gets or sets` makes more sense.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:502 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>           "description": "Gets or sets the module properties."

nit: remove `or sets` (below as well) as `module` is returned as response of GET/PUT/PATCH, and not used as inputs.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:530 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>           "description": "Gets or sets the module create properties."

nit: lets remove `Gets` from the docs for the reason explained below

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:385 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>           "description": "Gets or sets a list of modules."

nit: pls fix.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:321 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>           "description": "Gets or sets the error code."

nit: pls fix (remove sets)

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:304 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>         "contentLink": {

queried about this internally. It will be good if we can add proper mutability notations to indicate that it is only a PUTable property using

x-ms-muatbility: [
   "create",
   "update"
]

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:369 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by pcsin** [comment]
>         "operationId": "PowerShell72Module_Update",

Feedback in such cases where PATCH is to support tags updates is to update the operationId so it reflects it is a tag-only update operation.

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:197 in 0f85965. [](commit_id = 0f85965d7a3a335b874621b07ded822a6f7bd672, deletion_comment = False)

**Comment by mahakjain314** [comment]
Fixed all of the comments regarding description to change from Gets or Sets to only Sets/Gets wherever applicable.

**Comment by mahakjain314** [comment]
> > ```
> >     "contentLink": {
> > ```
> 
> queried about this internally. It will be good if we can add proper mutability notations to indicate that it is only a PUTable property using
> 
> x-ms-muatbility: [ "create", "update" ]
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:369 in [0f85965](https://github.com/Azure/azure-rest-api-specs/commit/0f85965d7a3a335b874621b07ded822a6f7bd672). [](commit_id = [0f85965](https://github.com/Azure/azure-rest-api-specs/commit/0f85965d7a3a335b874621b07ded822a6f7bd672), deletion_comment = False)

Added the notation as suggested.

**Comment by mahakjain314** [comment]
> > ```
> >     "operationId": "PowerShell72Module_Update",
> > ```
> 
> Feedback in such cases where PATCH is to support tags updates is to update the operationId so it reflects it is a tag-only update operation.
> 
> Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:197 in [0f85965](https://github.com/Azure/azure-rest-api-specs/commit/0f85965d7a3a335b874621b07ded822a6f7bd672). [](commit_id = [0f85965](https://github.com/Azure/azure-rest-api-specs/commit/0f85965d7a3a335b874621b07ded822a6f7bd672), deletion_comment = False)

ContentLink can also be updated not just tags. The two properties - tags and ContentLink can be patched individually or together. 
Patching the ContentLink reflects by update in properties like version, size etc.

**Comment by pcsin** [comment]
>       "x-ms-muatbility": [

spell check. "x-ms-mutability"

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:364 in 8bfd5c3. [](commit_id = 8bfd5c31f49845e89df3b983a23843767f7d67c0, deletion_comment = False)

**Comment by pcsin** [comment]
Pls refer for discussion

[https://teams.microsoft.com/l/message/19:78cd6c72c41240dc84a36523af60b88a@thread.tacv2/1697728539821?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=0cab4ce9-7691-42ae-82e3-460d4346a710&parentMessageId=1697728539821&teamName=ARM%20API%20Reviewers&channelName=ARM%20swagger%20reviews&createdTime=1697728539821](https://teams.microsoft.com/l/message/19:78cd6c72c41240dc84a36523af60b88a@thread.tacv2/1697728539821?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=0cab4ce9-7691-42ae-82e3-460d4346a710&parentMessageId=1697728539821&teamName=ARM%20API%20Reviewers&channelName=ARM%20swagger%20reviews&createdTime=1697728539821)

---
In reply to: [1757403137](https://github.com/Azure/azure-rest-api-specs/pull/26057#issuecomment-1757403137) [](http://example.com/codeflow?ancestors=1757403137)

**Review by pcsin — APPROVED** [review]
:shipit:

**Comment by rkmanda** [comment]
>               "$ref": "#/definitions/ModuleCreateOrUpdateParameters"

why not Module?

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:161 in 210e0fb. [](commit_id = 210e0fb4bbb3aa63ec3e3765e2d4839cbd149d51, deletion_comment = False)

**Comment by rkmanda** [comment]
>               "$ref": "#/definitions/ModuleCreateOrUpdateParameters"

[https://github.com/Azure/azure-rest-api-specs/pull/26057/checks?check_run_id=18115124058](https://github.com/Azure/azure-rest-api-specs/pull/26057/checks?check_run_id=18115124058)

---
In reply to: [1783365617](https://github.com/Azure/azure-rest-api-specs/pull/26057#issuecomment-1783365617) [](http://example.com/codeflow?ancestors=1783365617)

---
Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:161 in 210e0fb. [](commit_id = 210e0fb4bbb3aa63ec3e3765e2d4839cbd149d51, deletion_comment = False)

**Comment by rkmanda** [comment]
> > > ```
> > >           "$ref": "#/definitions/ModuleCreateOrUpdateParameters"
> > > ```
> > 
> > 
> > why not Module?
> > Refers to: specification/automation/resource-manager/Microsoft.Automation/stable/2023-11-01/powershell72Module.json:161 in [210e0fb](https://github.com/Azure/azure-rest-api-specs/commit/210e0fb4bbb3aa63ec3e3765e2d4839cbd149d51). [](commit_id = [210e0fb](https://github.com/Azure/azure-rest-api-specs/commit/210e0fb4bbb3aa63ec3e3765e2d4839cbd149d51), deletion_comment = False)
> 
> For Module resource, we need file uploads (currently taking a short lived blob content uri) in PUT request, so couldn't make PUT contract same as GET contract. PUT has only one property for contentURI and which we don't return as part of GET response.
> 
> **GET response has many read-only properties which are extracted in the import process.**
> 
> Also, we have 4 existing APIs which adhere to the same contract with routes /modules /powershell7Modules /python2Packages /python3Packages.
> 
> The backend implementation and contracts of all these APIs are same.
> 
> This is probably the last time we are adding this route. We have done a major redesign where we will not have to introduce new route for every runtime version we introduce. Refer to this PR for details - #23988

Its an anti-pattern to have the response of the GET be different from the request of the PUT. This is because usually clients use a GET-PUT semantics for replacing their resources and having the schemas different makes it difficult for client implementations. However, given you already have similar patterns earlier, we can approve this for now.

**Comment by rkmanda** [comment]
Please add a suppression for [PutRequestResponseSchemeArm](https://github.com/Azure/azure-openapi-validator/blob/main/docs/put-request-response-scheme-arm.md)

using the guidance provided below
https://github.com/Azure/autorest/blob/main/docs/generate/suppress-warnings.md

**Comment by mahakjain314** [comment]
> Please add a suppression for [PutRequestResponseSchemeArm](https://github.com/Azure/azure-openapi-validator/blob/main/docs/put-request-response-scheme-arm.md)
> 
> using the guidance provided below https://github.com/Azure/autorest/blob/main/docs/generate/suppress-warnings.md

Added the suppression rule as suggested.

**Review by rkmanda — APPROVED** [review]
:shipit:

File: specification/automation/resource-manager/Microsoft.Automation/common/v5/definitions.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    "AutomationAccountNameParameter": {
+      "name": "automationAccountName",
+      "description": "The name of the automation account.",
+      "type": "string",
+      "required": true,
+      "in": "path",
+      "x-ms-parameter-location": "method",
+      "pattern": "^[a-zA-Z][a-zA-Z-_0-9]*$"
+    },
+    "ModuleNameParameter": {
```
Comment (Alancere) [inline specification/automation/resource-manager/Microsoft.Automation/common/v5/definitions.json:37]: Hi @sushil490023, please add "x-ms-parameter-location": "method"

**Comment by rkmanda** [comment]
You are at step 3 - Pl get a breaking change approval Approved-SdkBreakingChange-Go and then request a merge again

**Pull request merged**